src/dnn/nnlib.jl

```diff
@@ -59,25 +59,50 @@ end
 
 # Convolution
 
+const convworkspace = [CuVector{UInt8}(1)]
+
+getconvworkspace(bytes) =
+  length(convworkspace[]) ≥ bytes ?
+    convworkspace[] :
+    (convworkspace[] = CuVector{UInt8}(bytes))
+
 function conv!(y::A, x::A, w::A;
                pad = 0, stride = 1, mode = 0,
-               alpha = 1, dilation = 1) where A<:CuArray{<:CUDNNFloat}
+               alpha = 1, dilation = 1, workspace = nothing,
+               workspaceSizeInBytes = 0, algo = 0) where A<:CuArray{<:CUDNNFloat}
   all(x -> x == 1, dilation) || error("Only dilation = 1 is supported in CuArrays")
-  cudnnConvolutionForward(y, x, w, padding=pad, stride=stride, mode=mode, alpha=alpha)
+  if workspace == nothing
+    workspaceSizeInBytes = cudnnGetConvolutionForwardWorkspaceSize(y, x, w, padding = pad, stride = stride, algo = algo, mode = mode)
+    workspace = workspaceSizeInBytes != 0 ? getconvworkspace(workspaceSizeInBytes) : workspace
+  end
+  cudnnConvolutionForward(y, x, w, padding=pad, stride=stride, mode=mode, alpha=alpha, algo = algo,
+                          workSpace=workspace, workSpaceSizeInBytes=workspaceSizeInBytes)
 end
 
 function ∇conv_filter!(dw::A, dy::A, x::A, w::A;
                        pad = 0, stride = 1, mode = 0,
-                       alpha = 1, dilation = 1) where A<:CuArray{<:CUDNNFloat}
+                       alpha = 1, dilation = 1, workspace = nothing,
+                       workSpacesizeInBytes = 0, algo = 0) where A<:CuArray{<:CUDNNFloat}
   all(x -> x == 1, dilation) || error("Only dilation = 1 is supported in CuArrays")
-  cudnnConvolutionBackwardFilter(dw, x, w, dy, padding=pad, stride=stride, mode=mode, alpha=alpha)
+  if workspace == nothing
+    workspaceSizeInBytes = cudnnGetConvolutionBackwardFilterWorkspaceSize(dw, x, w, dy, padding = pad, stride = stride, algo = algo, mode = mode)
+    workspace = workspaceSizeInBytes != 0 ? getconvworkspace(workspaceSizeInBytes) : workspace
+  end
+  cudnnConvolutionBackwardFilter(dw, x, w, dy, padding=pad, stride=stride, mode=mode, alpha=alpha, algo = algo,
+                                 workSpace=workspace, workSpaceSizeInBytes=workspaceSizeInBytes)
 end
 
 function ∇conv_data!(dx::A, dy::A, x::A, w::A;
                      pad = 0, stride = 1, mode = 0,
-                     alpha = 1, dilation = 1) where A<:CuArray{<:CUDNNFloat}
+                     alpha = 1, dilation = 1, workspace = nothing,
+                     workspaceSizeInBytes = 0, algo = 0) where A<:CuArray{<:CUDNNFloat}
```

Comment:
In these high-level APIs, is there a particular reason for having two keyword arguments for the workspace, rather than just doing passing a `Union{CuVector,Nothing}`?

Reply:
Nothing as such. Can simply do as suggested.